### ISSUE TITLE
Provisioning on CoreOS with sudo access with Azure Builder

### ISSUE THREAD

**Issue by luhkevin** [mitchellh/packer#4245]
packer version: 0.10.1
Host platform: Ubuntu 16.04
Debug log output: https://gist.github.com/luhkevin/2db9d6940160d3ac91f199fe031918d3
Example template and scripts:
```
{
  "builders": [
    {
      "type": "azure-arm",
      "tenant_id": "[REMOVED]",
      "subscription_id": "[REMOVED",
      "client_id": "[REMOVED]",
      "client_secret": "[REMOVED]",
      "resource_group_name": "netsil-builder",
      "storage_account": "netsilimages",
      "capture_container_name": "myimages",
      "capture_name_prefix": "NETSIL",
      "image_publisher": "CoreOS",
      "image_offer": "CoreOS",
      "image_sku": "Stable",
      "location": "westus2",
      "os_type": "Linux",
      "ssh_username": "core"
    }
  ],
  "provisioners": [
    {
      "inline":["whoami"],
      "type": "shell"
    },
    {
      "inline":["cat /etc/*release"],
      "type": "shell"
    },
    {
      "inline":["env"],
      "type": "shell"
    },
    {
      "execute_command": "chmod +x {{ .Path }}; {{ .Vars }} sudo -E sh '{{ .Path }}'",
      "inline":["/usr/share/oem/python/bin/python -u /usr/share/oem/bin/waagent -force -deprovision+user && export HISTSIZE=0 && sync"],
      "inline_shebang": "/bin/sh -x",
      "type": "shell"
    }
  ]
}
```

I'm trying out the packer azure builder with CoreOS, but the provisioner isn't working regarding sudo access. The details are in the log output, but essentially I'm seeing the message:

```
2016/12/03 07:01:07 ui:     azure-arm:
2016/12/03 07:01:07 packer: 2016/12/03 07:01:07 [INFO] 0 bytes written for 'stdout'
2016/12/03 07:01:07 packer: 2016/12/03 07:01:07 [INFO] 307 bytes written for 'stderr'
2016/12/03 07:01:07 packer: 2016/12/03 07:01:07 [INFO] RPC client: Communicator ended with: 1
2016/12/03 07:01:07 ui:     azure-arm: We trust you have received the usual lecture from the local System
2016/12/03 07:01:07 ui:     azure-arm: Administrator. It usually boils down to these three things:
2016/12/03 07:01:07 ui:     azure-arm:
2016/12/03 07:01:07 ui:     azure-arm: #1) Respect the privacy of others.
2016/12/03 07:01:07 ui:     azure-arm: #2) Think before you type.
2016/12/03 07:01:07 ui:     azure-arm: #3) With great power comes great responsibility.
2016/12/03 07:01:07 ui:     azure-arm:
2016/12/03 07:01:07 ui:     azure-arm: sudo: no tty present and no askpass program specified
```

After some googling, I've found that I should set `"ssh_pty": true`, but this doesn't do anything. It still prints the same thing and just waits for me to type in the password.

It's really strange because I created the same image on Azure manually and was able to do things like `sudo bash` with the core user without any password prompting, and without that output above.

For reference, the sudoers file on that image is:
```
## Default sudoers for CoreOS

## Pass LESSCHARSET through for systemd commands run through sudo that call less.
## See https://github.com/coreos/bugs/issues/365.
Defaults env_keep += "LESSCHARSET"

## enable root and wheel
root ALL=(ALL) ALL
%wheel ALL=(ALL) ALL

## enable passwordless access for sudo
%sudo ALL=(ALL) NOPASSWD: ALL

## the core user has no password by default
core ALL=(ALL) NOPASSWD: ALL

## Read drop-in files from /etc/sudoers.d
## (the '#' here does not indicate a comment)
#includedir /etc/sudoers.d
```

So the `core` user definitely has sudo access. The `whoami` command that I've run confirms that I'm the `core` user.

I've tried the example [here](https://github.com/mitchellh/packer/blob/master/examples/azure/ubuntu_quickstart.json) and it was working just fine. 

Maybe this is just a CoreOS/Azure issue?

**Comment by boumenot** [comment]
This is all dependent upon the image, and is not an Azure specific issue.  The example you used is a good one, but in my experience the Canonical images are generally the easiest to work with.  The Ubuntu example did not have to handle this case, but it was configured not to.  Checkout the [CentOS](https://github.com/mitchellh/packer/blob/master/examples/azure/centos.json) or [Debian](https://github.com/mitchellh/packer/blob/master/examples/azure/debian.json) example for a potential work around for your issue.  I have not tried CoreOS specifically, but the workaround may work for you.  Please let me know.

I need to update the Debian example to include the feedback from #4129, and I should add an example for CoreOS to make it easier for new users.

**Comment by luhkevin** [comment]
Do you mean set `"ssh_pass": null`, use `"execute_command": "echo '{{user `ssh_pass`}}' | {{ .Vars }} sudo -S -E sh '{{ .Path }}'",`, and set `"ssh_password": "{{user `ssh_pass`}}"` in the builder?

Because then it asks for a password. I'm getting:
`* required variable not set: ssh_pass`

Am I supposed to set a password for the coreos image? It does not have one by default.

**Comment by boumenot** [comment]
See the example for details, but define a variable called ssh_pass, and set the value via a user variable.  There's a null value in the examples to make it clear that one needs to be defined, but perhaps that isn't clear.

> packer -var ssh_pass="Password1234"

**Comment by luhkevin** [comment]
I'm still a little bit confused. Does this workaround actually try to authenticate with the CentOS/Debian images with a password, meaning that I would have to create a custom CoreOS image with a password? 

Or is the workaround simply passing these images some dummy password, that's not actually used for any sort of authentication?

**Comment by luhkevin** [comment]
Actually, it looks like using a dummy password worked, so thanks for the help on that!
Unfortunately, I was also trying to use an ansible provisioner which requires SSH access in a similar way, and that is still bothering me about a password. It's returning in stdout:

```
sudo: a password is required
```

I'll try the same approach for ansible -- pass it a dummy password -- and let you know my results.

**Comment by boumenot** [comment]
w00t!  Let me know if it works out for you or not.

When the Azure builder creates your VM it assign it a random password in the ARM template.  The template is deployed to create the VM, and the password flows through to the deployed VM.  The password is set by guest agent (or maybe cloud-init?!?).  This password is never exposed because it is a random value that you generally do not care about.  (If you enable debug mode the password will be printed to STDOUT.)  You can override the password by setting a value for ssh_password.  The only time you need the password is in cases where you need to echo it into sudo to elevate because the distro's configuration requires it.

> The term ssh_password is a misnomer because it is not applicable to Windows.  ssh_password means admin password to me, but it is called ssh_password.

It sounds like you're unblocked, so I'm closing the issue.

**Comment by luhkevin** [comment]
I managed to get around the password issue with ansible. I just had to pass 
`"extra_arguments": [ "--extra-vars", "ansible_become_pass=Password1234" ],` and set `sudo_flags=-H -S` in ansible.cfg. 

But there was another ansible issue regarding file copying, which I'll make another issue for.

**Comment by luhkevin** [comment]
Just following up on that ansible issue I was having -- it was that sftp error, and I simply had to upgrade my packer version and use scp instead. It's fully working now, thanks for all the help!

**Comment by boumenot** [comment]
Thanks for debugging this, and sharing with the community at large.  Glad to hear it works for you!

**Comment by luhkevin** [comment]
No problem. And actually, I had a quick follow-up question. The random password that gets assigned, does it actually get shipped with the image that I'm building? From what I can tell, you're saying it gets set at launch-time, and not build-time, but I just wanted to clarify.

**Comment by boumenot** [comment]
If you executed the deprovision process at the end of the build I would expect the user account to be deleted, which should remove any remnants.

Perhaps overkill, but ...

If you are really concerned about the password be recovered you will need to take more drastic measures, such as securely deleting it.  Data from VHDs can be recovered using the various un-delete or security forensic tools.

**Comment by luhkevin** [comment]
I see, the password is associated with the user I'm using to build it with. In my case, the `core` user?
And just in case, how would I go about securely deleting it?

**Comment by boumenot** [comment]
Do a search for "linux secure delete."  Look for tools like shred and sfill.
